sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnnecessaryUsings.cs

```diff
@@ -147,6 +149,35 @@ public override void VisitAwaitExpression(AwaitExpressionSyntax node)
                 base.VisitAwaitExpression(node);
             }
 
+            /// <summary>
+            /// LINQ Query Syntax do not use symbols from the 'System.Linq' namespace directly, but the using directive is
+            /// still necessary to use the Query Syntax form.
+            /// </summary>
+            public override void VisitQueryExpression(QueryExpressionSyntax node)
+            {
+                if (!this.linqQueryVisited && TryGetSystemLinkNamespace(out var systemLinqNamespaceSymbol))
+                {
+                    this.necessaryNamespaces.Add(systemLinqNamespaceSymbol);
+                }
+                this.linqQueryVisited = true;
+                base.VisitQueryExpression(node);
+            }
+
+            private bool TryGetSystemLinkNamespace(out INamespaceSymbol systemLinqNamespace)
+            {
+                foreach (var usingDirective in this.usingDirectivesFromParent)
+                {
+                    if (this.context.SemanticModel.GetSymbolInfo(usingDirective.Name).Symbol is INamespaceSymbol namespaceSymbol &&
+                        namespaceSymbol.ToDisplayString() == "System.Linq")
```

Comment:
Would this work even if the code being analyzed was written as such?

```csharp
namespace System {
  using Linq;

  ...
}
```

Reply:
Good point.
The code snippet you mentioned 
```
if (this.context.SemanticModel.GetSymbolInfo(usingDirective.Name).Symbol is INamespaceSymbol namespaceSymbol &&
    namespaceSymbol.ToDisplayString() == "System.Linq")
```
would actually work, as it relies on the display name of the symbol. So it is correctly resolved as `System.Linq` when `Linq` is nested inside `System` namespace.
However, for performance reason, that part of the code was changed and the newest version (currently on master) does not work in that case.
I've created a new PR to revert to the version relying on symbol: (https://github.com/SonarSource/sonar-dotnet/pull/2818)